proknow-sdk/Audit/AuditPage.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System.Collections.Generic;
+using System.Text.Json.Serialization;
+using System.Threading.Tasks;
+using System.Text.Json;
+using System.Net.Http;
+using System.Text;
+using ProKnow.Exceptions;
+
+namespace ProKnow.Audit
+{
+    /// <summary>
+    /// Represents a collection of audit log items
+    /// </summary>
+    public class AuditPage
+    {
+        private ProKnowApi _proKnow;
+        private FilterParametersExtended _filterParameters;
+
+        private JsonSerializerOptions _serializerOptions = new JsonSerializerOptions
+        {
+            IgnoreNullValues = true,
+        };
+
+        /// <summary>
+        /// The total number of log entries
+        /// </summary>
+        [JsonPropertyName("total")]
+        public uint Total { get; set; }
+
+        /// <summary>
+        /// The audit log items returned
+        /// </summary>
+        [JsonPropertyName("items")]
+        public List<AuditItem> Items { get; set; }
+
+        /// <summary>
+        /// Finishes initialization of object after deserialization from JSON
+        /// </summary>
+        /// <param name="proKnow">Root object for interfacing with the ProKnow API</param>
+        /// <param name="parameters">The filter paramters used in API call</param>
+        internal void PostProcessDeserialization(ProKnowApi proKnow, FilterParametersExtended parameters)
+        {
+            _proKnow = proKnow;
+            _filterParameters = parameters;
+        }
+
+        /// <summary>
+        /// Gets next page of audit logs asynchronously
+        /// </summary>
+        /// <returns>The next page of audit logs</returns>
+        /// <example>This example shows how to get the next page of audit logs:
+        /// <code>
+        /// using ProKnow;
+        /// using System.Threading.Tasks;
+        ///
+        /// var pk = new ProKnowApi("https://example.proknow.com", "./credentials.json");
+        /// var page = await _proKnow.Audit.Query(filterParams);
+        /// var nextPage = await page.Next();
+        /// </code>
+        /// </example>
+        public async Task<AuditPage> Next()
+        {
+            if (this._filterParameters == null)
+            {
+                throw new ProKnowException("Must call Query first");
+            } else if (this._filterParameters.FirstId == null)
+            {
+                throw new ProKnowException("Must call Query first");
+            }
+
+            ++this._filterParameters.PageNumber;
```

Comment:
Consider this code:

```c#
using ProKnow;
using System.Threading.Tasks;

var pk = new ProKnowApi("https://example.proknow.com", "./credentials.json");
var page1 = await _proKnow.Audit.Query(filterParams);
var page2 = await page1.Next();
var page2Again = await page1.Next();
```

With the way it's currently implemented, I think the variable `page2Again` would actually have page 3 data. To fix this, I think you would have to copy this._filterParameters and then increment the page number before passing it into the Serialize method.

Reply:
Fixed, and updated test case to cover this scenario